src/QMCWaveFunctions/Jastrow/J1Spin.h

```diff
@@ -0,0 +1,854 @@
+   * @param iat particle index
+   */
+  GradType evalGrad(ParticleSet& P, int iat) override { return GradType(Grad[iat]); }
+
+  /** compute the gradient during particle-by-particle update
+   * @param P quantum particleset
+   * @param iat particle index
+   *
+   * Using getTempDists(). curAt, curGrad and curLap are computed.
+   */
+  PsiValueType ratioGrad(ParticleSet& P, int iat, GradType& grad_iat) override
+  {
+    UpdateMode = ORB_PBYP_PARTIAL;
+
+    computeU3(P, iat, P.getDistTable(myTableID).getTempDists());
+    curLap = accumulateGL(dU.data(), d2U.data(), P.getDistTable(myTableID).getTempDispls(), curGrad);
+    curAt  = simd::accumulate_n(U.data(), Nions, valT());
+    grad_iat += curGrad;
+    return std::exp(static_cast<PsiValueType>(Vat[iat] - curAt));
+  }
+
+  /** Rejected move. Nothing to do */
+  inline void restore(int iat) override {}
+
+  /** Accpted move. Update Vat[iat],Grad[iat] and Lap[iat] */
+  void acceptMove(ParticleSet& P, int iat, bool safe_to_delay = false) override
+  {
+    if (UpdateMode == ORB_PBYP_RATIO)
+    {
+      computeU3(P, iat, P.getDistTable(myTableID).getTempDists());
+      curLap = accumulateGL(dU.data(), d2U.data(), P.getDistTable(myTableID).getTempDispls(), curGrad);
+    }
+
+    LogValue += Vat[iat] - curAt;
+    Vat[iat]  = curAt;
+    Grad[iat] = curGrad;
+    Lap[iat]  = curLap;
+  }
+
+
+  inline void registerData(ParticleSet& P, WFBufferType& buf) override
+  {
+    if (Bytes_in_WFBuffer == 0)
+    {
+      Bytes_in_WFBuffer = buf.current();
+      buf.add(Vat.begin(), Vat.end());
+      buf.add(Grad.begin(), Grad.end());
+      buf.add(Lap.begin(), Lap.end());
+      Bytes_in_WFBuffer = buf.current() - Bytes_in_WFBuffer;
+      // free local space
+      Vat.free();
+      Grad.free();
+      Lap.free();
+    }
+    else
+    {
+      buf.forward(Bytes_in_WFBuffer);
+    }
+  }
+
+  inline LogValueType updateBuffer(ParticleSet& P, WFBufferType& buf, bool fromscratch = false) override
+  {
+    evaluateGL(P, P.G, P.L, false);
+    buf.forward(Bytes_in_WFBuffer);
+    return LogValue;
+  }
+
+  inline void copyFromBuffer(ParticleSet& P, WFBufferType& buf) override
+  {
+    Vat.attachReference(buf.lendReference<valT>(Nelec), Nelec);
+    Grad.attachReference(buf.lendReference<posT>(Nelec), Nelec);
+    Lap.attachReference(buf.lendReference<valT>(Nelec), Nelec);
+  }
+
+  inline void setVars(const opt_variables_type& vars)
+  {
+    NumVars = vars.size();
+    if (NumVars == 0)
+      return;
+    myVars = vars;
+    dLogPsi.resize(NumVars);
+    gradLogPsi.resize(NumVars, 0);
+    lapLogPsi.resize(NumVars, 0);
+    for (int i = 0; i < NumVars; ++i)
+    {
+      gradLogPsi[i] = new WavefunctionFirstDerivativeType(Nelec);
+      lapLogPsi[i]  = new WavefunctionSecondDerivativeType(Nelec);
+    }
+  }
+
+
+  std::unique_ptr<WaveFunctionComponent> makeClone(ParticleSet& tqp) const override
+  {
+    auto j1copy         = std::make_unique<J1Spin<FT>>(myName, Ions, tqp);
```

Comment:
please move all the contents of makeClone into a copy constructor and makeClone should just invoke the copy constructor.

Reply:
Done